src/modules/sensors/vehicle_air_data/VehicleAirData.cpp

```diff
@@ -0,0 +1,280 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "VehicleAirData.hpp"
+
+#include <px4_platform_common/log.h>
+#include <lib/ecl/geo/geo.h>
+
+using namespace matrix;
+using namespace time_literals;
+
+static constexpr uint32_t SENSOR_TIMEOUT{300_ms};
+
+VehicleAirData::VehicleAirData() :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::att_pos_ctrl)
+{
+	_voter.set_timeout(SENSOR_TIMEOUT);
+}
+
+VehicleAirData::~VehicleAirData()
+{
+	Stop();
+
+	perf_free(_cycle_perf);
+}
+
+bool VehicleAirData::Start()
+{
+	ScheduleNow();
+
+	return true;
+}
+
+void VehicleAirData::Stop()
+{
+	Deinit();
+
+	// clear all registered callbacks
+	for (auto &sub : _sensor_sub) {
+		sub.unregisterCallback();
+	}
+}
+
+void VehicleAirData::SensorCorrectionsUpdate()
+{
+	// check if the selected sensor has updated
+	sensor_correction_s corrections;
+
+	if (_sensor_correction_sub.update(&corrections)) {
+		for (int baro_index = 0; baro_index < MAX_SENSOR_COUNT; baro_index++) {
+			const sensor_baro_s &baro = _last_data[baro_index];
+
+			// find sensor (by device id) in sensor_correction
+			if (_sensor_correction_index[baro_index] < 0) {
+				for (int correction_index = 0; correction_index < MAX_SENSOR_COUNT; correction_index++) {
+					if ((baro.device_id > 0) && (corrections.baro_device_ids[correction_index] == baro.device_id)) {
+						_sensor_correction_index[baro_index] = correction_index;
+					}
+				}
+			}
+
+			switch (_sensor_correction_index[baro_index]) {
+			case 0:
+				_offset[baro_index] = corrections.baro_offset_0;
+				_scale[baro_index] = corrections.baro_scale_0;
+				break;
+
+			case 1:
+				_offset[baro_index] = corrections.baro_offset_1;
+				_scale[baro_index] = corrections.baro_scale_1;
+				break;
+
+			case 2:
+				_offset[baro_index] = corrections.baro_offset_2;
+				_scale[baro_index] = corrections.baro_scale_2;
+				break;
+
+			default:
+				_offset[baro_index] = 0.f;
+				_scale[baro_index] = 1.f;
+			}
+		}
+	}
+}
+
+void VehicleAirData::ParametersUpdate()
+{
+	// Check if parameters have changed
+	if (_params_sub.updated()) {
+		// clear update
+		parameter_update_s param_update;
+		_params_sub.copy(&param_update);
+
+		updateParams();
+	}
+}
+
+void VehicleAirData::Run()
+{
+	perf_begin(_cycle_perf);
+
+	bool updated[MAX_SENSOR_COUNT] {};
+
+	for (int uorb_index = 0; uorb_index < MAX_SENSOR_COUNT; uorb_index++) {
+
+		if (!_advertised[uorb_index]) {
+			// use data's timestamp to throttle advertisement checks
+			if (hrt_elapsed_time(&_last_data[uorb_index].timestamp) > 1_s) {
+				if (_sensor_sub[uorb_index].advertised()) {
+					if (uorb_index > 0) {
+						/* the first always exists, but for each further sensor, add a new validator */
+						if (!_voter.add_new_validator()) {
+							PX4_ERR("failed to add validator for sensor_baro:%i", uorb_index);
```

Comment:
What happens if this failed?

Reply:
Pushing in new data later would fail silently. In general we don't handle OOM gracefully all over the system, and I'm not sure if we ever will.

For the data validator in particular it probably makes sense to move this back PX4/Firmware side and not have most of the dynamic allocations in the first place. I'd like to get to the point where we effectively don't have any memory allocations past initial startup.